### PR TITLE
TYPING: Improve ``np.sum`` and ``np.mean`` return types with given ``dtype``

### LINKED ISSUE #27638: TYP: improve `np.sum` and `np.mean`'s return types with given `dtype`
### Describe the issue:

In some cases, `sum` and `mean` return `Any` according to the typing information. I propose the following improvements:

### Reproduce the code example:

```diff
diff --git a/numpy/_core/fromnumeric.pyi b/numpy/_core/fromnumeric.pyi
index 32f8a06f7b..185b9472ec 100644
--- a/numpy/_core/fromnumeric.pyi
+++ b/numpy/_core/fromnumeric.pyi
@@ -597,15 +597,25 @@ def sum(
     where: _ArrayLikeBool_co = ...,
 ) -> _SCT: ...
 @overload
+def sum(
+    a: _ArrayLike,
+    axis: None = ...,
+    dtype: _DTypeLike[_SCT] = ...,
+    out: None = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike_co = ...,
+    where: _ArrayLikeBool_co = ...,
+) -> _SCT: ...
+@overload
 def sum(
     a: ArrayLike,
     axis: None | _ShapeLike = ...,
-    dtype: DTypeLike = ...,
+    dtype: _DTypeLike[_SCT] = ...,
     out: None  = ...,
     keepdims: bool = ...,
     initial: _NumberLike_co = ...,
     where: _ArrayLikeBool_co = ...,
-) -> Any: ...
+) -> _SCT | NDArray[_SCT]: ...
 @overload
 def sum(
     a: ArrayLike,
@@ -1210,12 +1220,12 @@ def mean(
 def mean(
     a: _ArrayLikeComplex_co | _ArrayLikeObject_co,
     axis: None | _ShapeLike = ...,
-    dtype: DTypeLike = ...,
+    dtype: _DTypeLike[_SCT] = ...,
     out: None = ...,
     keepdims: bool = ...,
     *,
     where: _ArrayLikeBool_co = ...,
-) -> Any: ...
+) -> _SCT | NDArray[_SCT]: ...
 @overload
 def mean(
     a: _ArrayLikeComplex_co | _ArrayLikeObject_co,
```

### Error message:

_No response_

### Python and NumPy Versions:

2.1.2
3.13.0 (main, Oct 16 2024, 00:33:24) [MSC v.1929 64 bit (AMD64)]

### Type-checker version and settings:

mypy

### Additional typing packages.

_No response_

### PR DESCRIPTION
**Pull request by bersbersbers**
Close #27638

### CODE REVIEW

**Review by jorenham — CHANGES_REQUESTED** [review]
Thanks for this!

I left some suggestions on how you could improve this. These are all related to issues that are actually quite common in the stubs. 
But even so, I think that if we can avoid it, then we shouldn't repeat the mistakes of others (or yourself), as common as they might be 🤷🏻 

Also I think it would be helpful if you could add a quick test or two to verify that this actually resolves the issue, so e.g. in `numpy/typing/tests/data/reveal/fromnumeric.pyi`.

**Comment by bersbersbers** [comment]
@jorenham thanks for the instructions - tried to account for all of them, let me know what you think.

File: numpy/_core/fromnumeric.pyi [hunk 1]
```diff
@@ -595,8 +605,50 @@ def sum(
     keepdims: bool = ...,
     initial: _NumberLike_co = ...,
     where: _ArrayLikeBool_co = ...,
+) -> _SCT | NDArray[_SCT]: ...
+@overload
+def sum(
+    a: _ArrayLike,
```
Comment (jorenham) [inline numpy/_core/fromnumeric.pyi:611]:
```suggestion
    a: ArrayLike,
```

File: numpy/_core/fromnumeric.pyi [hunk 2]
```diff
@@ -595,8 +605,50 @@ def sum(
...
+    axis: None,
+    dtype: _DTypeLike[_SCT],
+    out: None = ...,
+    keepdims: Literal[False] = ...,
+    initial: _NumberLike_co = ...,
+    where: _ArrayLikeBool_co = ...,
+) -> _SCT: ...
+@overload
+def sum(
+    a: _ArrayLike,
```
Comment (jorenham) [inline numpy/_core/fromnumeric.pyi:621]:
```suggestion
    a: ArrayLike,
```

File: numpy/_core/fromnumeric.pyi [hunk 3]
```diff
@@ -595,8 +605,50 @@ def sum(
...
+    out: None  = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike_co = ...,
+    where: _ArrayLikeBool_co = ...,
+) -> _SCT | NDArray[_SCT]: ...
+@overload
+def sum(
+    a: ArrayLike,
+    *,
+    axis: None | _ShapeLike = ...,
```
Comment (jorenham) [inline numpy/_core/fromnumeric.pyi:644]:
```suggestion
    axis: None | _ShapeLike = ...,
    *,
```

File: numpy/_core/fromnumeric.pyi [hunk 4]
```diff
@@ -595,8 +605,50 @@ def sum(
...
+    out: None = ...,
+    keepdims: Literal[False] = ...,
+    initial: _NumberLike_co = ...,
+    where: _ArrayLikeBool_co = ...,
+) -> _SCT: ...
+@overload
+def sum(
+    a: _ArrayLike,
+    *,
+    axis: None = ...,
```
Comment (jorenham) [inline numpy/_core/fromnumeric.pyi:623]:
```suggestion
    axis: None = ...,
    *,
```

File: numpy/_core/fromnumeric.pyi [hunk 5]
```diff
@@ -1207,6 +1259,26 @@ def mean(
...
+    out: None = ...,
+    keepdims: bool = ...,
+    *,
+    where: _ArrayLikeBool_co = ...,
+) -> _SCT | NDArray[_SCT]: ...
+@overload
+def mean(
+    a: _ArrayLikeComplex_co | _ArrayLikeObject_co,
+    *,
+    axis: None = ...,
```
Comment (jorenham) [inline numpy/_core/fromnumeric.pyi:1275]:
```suggestion
    axis: None = ...,
    *,
```

**Review by jorenham — CHANGES_REQUESTED** [review]
I like the systematically thorough tests 👌🏻.

I left a couple of minor suggestions. 
If you could address these, and then squash the commits with a nice commit message that starts with `TYP: ` (see the [docs](https://numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message) for details), then I think we can merge this :)

**Comment by jorenham** [comment]
did you mean to delete the fork 🤔 ?

**Pull request closed without merging**